src/main/java/org/twilmes/sql/gremlin/processor/SingleQueryExecutor.java

```diff
@@ -381,5 +348,149 @@ SqlGremlinQueryResult convertResult(final List<Object> rows, final RelNode input
                 columnTypes.add((col == null || col.getType() == null) ? "string" : col.getType());
             }
         }
+
+        public void assertIsEmpty() {
+            synchronized (assertEmptyLock) {
+                if (currThread != null && blockingQueueRows.size() == 0) {
+                    currThread.interrupt();
+                }
+                isEmpty = true;
+            }
+        }
+
+        public void addResults(final List<List<Object>> rows) {
+            blockingQueueRows.addAll(rows);
+        }
+
+        public Object getResult() {
+            try {
+                synchronized (assertEmptyLock) {
+                    // pass current thread in, and interrupt in assertIsEmpty
+                    this.currThread = Thread.currentThread();
+                    if (isEmpty && blockingQueueRows.size() == 0) {
+                        return null;
+                    }
+                }
+                return this.blockingQueueRows.take();
+            } catch (final InterruptedException ignored) {
+                return null;
+            }
+        }
+    }
+
+    public class VertexPagination implements Runnable {
+        private RelNode input;
+        private RelNode parent;
+
+        VertexPagination(final RelNode input, final RelNode parent) {
+            this.input = input;
+            this.parent = parent;
+        }
+
+        @Override
+        public void run() {
+            while (traversal.hasNext()) {
+                final RelDataType rowType = input.getRowType();
+                final List<String> fieldNames = rowType.getFieldNames();
+                final List<Map<String, Object>> results = (List<Map<String, Object>>) traversal.next(pageSize);
+                final List<Object> rows = new ArrayList<>();
+                for (final Map<String, Object> map : results) {
+                    final Map<Object, Object> mapResult = (Map<Object, Object>) map.get(MAP_KEY);
```

Comment:
Just as a comment, when you have code like this in the future I would suggest trying to abstract some of this out since you have some repetition. Now I know I wrote it and left it like that, so my bad, and I am going to change it anyway, so don't bother here. My suggestion for abstracting would be making a single pagination function that takes a lambda that does this processing and then leave the rest the same. There's probably other ways of doing it though. But just a comment for the future - no need to actually make changes.

Reply:
Sounds good. I can make the change for abstraction when we merge your join implementations.